src/ASM/ASMmxBase.C

```diff
@@ -229,6 +229,70 @@ ASMmxBase::SurfaceVec ASMmxBase::establishBases(Go::SplineSurface* surf,
                                                                   u0,vg,XYZ1,ndim,
                                                                   false,XYZ1));
     geoBasis = 3;
+  } else if (type == SUBGRID) {
+    // basis1 should be one degree higher than basis2 and C^p-1 continuous
+    int ndim = surf->dimension();
+    result[1].reset(new Go::SplineSurface(*surf));
+    Go::SplineSurface tmp(*surf);
+    for (size_t i = 0; i < 2; ++i) {
+      RealArray extraKnots;
+      RealArray::const_iterator uit = tmp.basis(i).begin();
+      double ucurr, uprev = *(uit++);
+      while (uit != tmp.basis(i).end())
+      {
+        ucurr = *(uit++);
+        if (ucurr > uprev)
+          for (int i = 0; i < 1; i++)
+          {
+            double xi = (double)(i+1)/(double)(1+1);
+            extraKnots.push_back(ucurr*xi + uprev*(1.0-xi));
+          }
+        uprev = ucurr;
+      }
+      if (i == 0)
+        tmp.insertKnot_u(extraKnots);
+      else
+        tmp.insertKnot_v(extraKnots);
+    }
+    Go::BsplineBasis b1 = tmp.basis(0).extendedBasis(tmp.order_u()+1);
+    Go::BsplineBasis b2 = tmp.basis(1).extendedBasis(tmp.order_v()+1);
+
+    // Compute parameter values of the Greville points
+    size_t i;
+    RealArray ug(b1.numCoefs()), vg(b2.numCoefs());
+    for (i = 0; i < ug.size(); i++)
+      ug[i] = b1.grevilleParameter(i);
+    for (i = 0; i < vg.size(); i++)
+      vg[i] = b2.grevilleParameter(i);
+
+    if (surf->rational()) {
+      std::vector<double> rCoefs(tmp.rcoefs_begin(), tmp.rcoefs_end());
+
+      // we normally would set coefs as (x*w, y*w, w)
+      // however, gotools use this representation internally already.
+
+      // instance a Bspline surface in ndim+1
+      Go::SplineSurface surf2(tmp.basis(0), tmp.basis(1), rCoefs.begin(), ndim+1, false);
+
+      // interpolate the Bspline surface onto new basis
+      RealArray XYZ((ndim+1)*ug.size()*vg.size());
+      surf2.gridEvaluator(XYZ,ug,vg);
+      std::unique_ptr<Go::SplineSurface> surf3(Go::SurfaceInterpolator::regularInterpolation(b1,b2,ug,vg,XYZ,ndim+1,false,XYZ));
+
+      // new rational coefs are (x/w', y/w', w')
+      // apparently gotools will rescale coeffs on surface creation.
+      result[0].reset(new Go::SplineSurface(surf3->basis(0), surf3->basis(1), surf3->coefs_begin(), ndim, true));
+    } else {
+      RealArray XYZ(ndim*ug.size()*vg.size());
+      // Evaluate the spline surface at all points
+      tmp.gridEvaluator(XYZ,ug,vg);
+
+      // Project the coordinates onto the new basis (the 2nd XYZ is dummy here)
+      result[0].reset(Go::SurfaceInterpolator::regularInterpolation(b1,b2,
```

Comment:
What you are doing is a pure knot insertion scheme. Would it not be shorter to simply clone `*surf` followed by knot insertion on this object. This should compute all the controlpoints to their proper value without interpolation schemes

Reply:
probably. i just copied the other code and made adjustments, will look into it.